Automattic/wp-calypso
# Improve login redirects and step tracking in stepper

Related to:
* https://github.com/Automattic/wp-calypso/pull/91074
* https://github.com/Automattic/wp-calypso/pull/90711

Fixes https://github.com/Automattic/wp-calypso/issues/91243

## Proposed Changes

* This PR introduces some key concepts into the core Stepper framework to allow us to better handle logged out users and the redirects that stem from them. The key changes are as follows:
  - Stepper steps now have an optional `requiresLoggedInUser` flag that should be specified in `flow.useSteps()`
    * The PR includes `stepWithRequiredLogin()` and `stepsWithRequiredLogin()` helpers to set this flag easily
  - The pre-existing `StepRoute` component now checks whether the current step has a truthy `requiresLoggedInUser` value, and will not render any content if the user is also not logged in, as well as triggering a redirect to log the user in.
    * The login redirect relies on a new `useLoginUrlForFlow()` hook that builds on my explorations in https://github.com/Automattic/wp-calypso/pull/90711
  - The logic to trigger `calypso_signup_step_start` Tracks events has been moved from `FlowRenderer` to `StepRoute`, and we won't trigger the Tracks event if the `StepRoute` logic from above performs a login redirect for a step that requires a login

## Why are these changes being made?



* We've been having internal discussions around improvements to the way Stepper handles Tracks and logins, and one of the key items that was flagged is that the current implementation can send `calypso_signup_step_start` Tracks events _even when the user doesn't see the step_. This PR attempts to resolve this by "knowing" whether a specific step requires the user to be logged in, and performing a redirect when necessary, as well as skipping rendering and triggering a Tracks event when we know a redirect should occur.
  - For now, the new skip logic only occurs when the step is "opted in" to the new requirement, but it should be relatively easy to roll out the behaviour on a flow-by-flow basis.

## Testing Instructions



* Run this branch locally or via Calypso.live
* Open a window where you are not logged in
* Ensure you have some way of seeing Tracks events
* Navigate to any of the URLs at or below `/setup/site-migration`, including:
  - `/setup/site-migration/`
  - `/setup/site-migration/site-migration-identify`
  - `/setup/site-migration/site-migration-import-or-migrate`
  - `/setup/site-migration/site-migration-upgrade-plan`
  - `/setup/site-migration/site-migration-instructions-i2`
* Verify that you're redirected to the login page, and the redirect URL will return you to your original location. Also verify that no `calypso_signup_step_start` Tracks event was recorded for the `site-migration` flow.
* Add URL parameters to the URLs above, and verify that you are still redirected to login and these URL parameters are included in the redirect URL used by the login page. You should still see no `calypso_signup_step_start` Tracks events for the `site-migration` flow.
* Specify a valid `locale` URL parameter for some of the above URLs, and verify that you're redirected to the right login page with the locale in the path _and_ the redirect URL for the login page includes the relevant non-English locale.  You should still see no `calypso_signup_step_start` Tracks events for the `site-migration` flow.
* In a second window where you are logged in, work through the same set of URLs above and verify that you're seeing the `calypso_signup_step_start` event being fired correctly for the `site-migration` flow
* In the same logged-in window, work through some other flows and verify that you're seeing the `calypso_signup_step_start` event being fired correctly

## Pre-merge Checklist
